File: ios/ExpoAdaptiveGlassModule.swift
import ExpoModulesCore

public class ExpoAdaptiveGlassModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoAdaptiveGlass")

    Events("onStats")

    OnCreate {
      GlassCapabilityDetector.prewarm()
    }

    Function("getDeviceInfo") {
      GlassCapabilityDetector.deviceInfo()
    }

    OnStartObserving {
      DispatchQueue.main.async { [weak self] in
        GlassPerformanceMonitor.shared.startObserving { payload in
          self?.sendEvent("onStats", payload)
        }
      }
    }

    OnStopObserving {
      DispatchQueue.main.async { GlassPerformanceMonitor.shared.stopObserving() }
    }

    OnDestroy {
      DispatchQueue.main.async { GlassPerformanceMonitor.shared.stopObserving() }
    }

    View(ExpoAdaptiveGlassView.self) {
      Events("onInteractionStart", "onInteractionEnd", "onDragStart", "onDragEnd")

      Prop("surfaceId") { (view, value: String) in view.surfaceId = value }
      Prop("renderer") { (view, value: String) in view.renderer = value }
      Prop("quality") { (view, value: String) in view.quality = value }
      Prop("blur") { (view, value: Double) in view.blur = CGFloat(value) }
      Prop("refraction") { (view, value: Double) in view.refraction = CGFloat(value) }
      Prop("dynamicHighlights") { (view, value: Bool) in view.dynamicHighlights = value }
      Prop("opaque") { (view, value: Bool) in view.opaqueMaterial = value }
      Prop("reduceMotion") { (view, value: Bool) in view.reduceMotion = value }
      Prop("intensity") { (view, value: Double) in view.intensity = CGFloat(value) }
      Prop("tintColor") { (view, value: UIColor?) in view.glassTint = value }
      Prop("tintScheme") { (view, value: String) in view.tintScheme = value }
      Prop("cornerRadius") { (view, value: Double) in view.cornerRadius = CGFloat(value) }
      Prop("interactive") { (view, value: Bool) in view.interactive = value }
      Prop("draggable") { (view, value: Bool) in view.draggable = value }
      // Android only
      Prop("shaderQuality") { (_: ExpoAdaptiveGlassView, _: Int) in }

      OnViewDidUpdateProps { view in
        view.propsDidUpdate()
      }
    }
  }
}

File: android/src/main/java/expo/modules/adaptiveglass/ExpoAdaptiveGlassModule.kt
package expo.modules.adaptiveglass

import expo.modules.adaptiveglass.performance.GlassCapabilityDetector
import expo.modules.adaptiveglass.performance.GlassPerformanceMonitor
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoAdaptiveGlassModule : Module() {
  private val context
    get() = requireNotNull(appContext.reactContext) { "React context is not available" }

  override fun definition() = ModuleDefinition {
    Name("ExpoAdaptiveGlass")

    Events("onStats")

    Function("getDeviceInfo") {
      GlassCapabilityDetector.info(context)
    }

    OnStartObserving {
      GlassPerformanceMonitor.start(context, appContext.currentActivity) { sendEvent("onStats", it) }
    }

    OnStopObserving {
      GlassPerformanceMonitor.stop()
    }

    OnActivityEntersForeground {
      GlassPerformanceMonitor.setForeground(true, appContext.currentActivity)
    }

    OnActivityEntersBackground {
      GlassPerformanceMonitor.setForeground(false, null)
    }

    OnDestroy {
      GlassPerformanceMonitor.stop()
    }

    View(ExpoAdaptiveGlassView::class) {
      Events("onInteractionStart", "onInteractionEnd", "onDragStart", "onDragEnd", "onMorphEnd", "onMenuSelect", "onMenuDismiss", "onMenuLongPress")

      Prop("surfaceId") { view: ExpoAdaptiveGlassView, value: String -> view.surfaceId = value }
      Prop("renderer") { view: ExpoAdaptiveGlassView, value: String -> view.renderer = value }
      Prop("quality") { view: ExpoAdaptiveGlassView, value: String -> view.quality = value }
      Prop("blur") { view: ExpoAdaptiveGlassView, value: Double -> view.blur = value.toFloat() }
      Prop("refraction") { view: ExpoAdaptiveGlassView, value: Double -> view.refraction = value.toFloat() }
      Prop("dynamicHighlights") { view: ExpoAdaptiveGlassView, value: Boolean -> view.dynamicHighlights = value }
      Prop("shaderQuality") { view: ExpoAdaptiveGlassView, value: Int -> view.shaderQuality = value }
      Prop("opaque") { view: ExpoAdaptiveGlassView, value: Boolean -> view.opaque = value }
      Prop("reduceMotion") { view: ExpoAdaptiveGlassView, value: Boolean -> view.reduceMotion = value }
      Prop("intensity") { view: ExpoAdaptiveGlassView, value: Double -> view.intensity = value.toFloat() }
      Prop("clarity") { view: ExpoAdaptiveGlassView, value: Double -> view.clarity = value.toFloat() }
      Prop("tintColor") { view: ExpoAdaptiveGlassView, value: Int? -> view.tint = value }
      Prop("tintScheme") { view: ExpoAdaptiveGlassView, value: String -> view.tintScheme = value }
      Prop("cornerRadius") { view: ExpoAdaptiveGlassView, value: Double -> view.cornerRadiusDp = value.toFloat() }
      Prop("shadow") { view: ExpoAdaptiveGlassView, value: Double -> view.shadow = value.toFloat() }
      Prop("edgeColor") { view: ExpoAdaptiveGlassView, value: Int? -> view.edgeColor = value }
      Prop("edgeWidth") { view: ExpoAdaptiveGlassView, value: Double -> view.edgeWidth = value.toFloat() }
      Prop("edgeRefraction") { view: ExpoAdaptiveGlassView, value: Double -> view.edgeRefraction = value.toFloat() }
      Prop("interactive") { view: ExpoAdaptiveGlassView, value: Boolean -> view.interactive = value }
      Prop("draggable") { view: ExpoAdaptiveGlassView, value: Boolean -> view.draggable = value }
      Prop("morphRect") { view: ExpoAdaptiveGlassView, value: Map<String, Double>? -> view.morphRect = value }
      Prop("morphIndex") { view: ExpoAdaptiveGlassView, value: Int -> view.morphIndex = value }
      Prop("menuRows") { view: ExpoAdaptiveGlassView, value: List<List<Double>>? -> view.menuRows = value ?: emptyList() }
      Prop("menuTrigger") { view: ExpoAdaptiveGlassView, value: Boolean -> view.menuTrigger = value }

      OnViewDidUpdateProps { view: ExpoAdaptiveGlassView ->
        view.propsDidUpdate()
      }
    }
  }
}
